File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& arr) {
        // if (intervals.size() <= 1) return intervals;

    
        // sort(intervals.begin(), intervals.end());

        // vector<vector<int>> ans;
        // vector<int> temp = intervals[0]; 

        // for (const auto& interval : intervals) {
        //     if (interval[0] <= temp[1]) {
               
        //         temp[1] = max(temp[1], interval[1]);
        //     } else {
               
        //         ans.push_back(temp);
        //         temp = interval;
        //     }
        // }

        // ans.push_back(temp);
        // return ans;

         vector<vector<int>>ans;
        sort(arr.begin(),arr.end());
        ans.push_back({arr[0][0],arr[0][1]});
        for(int i=1;i<arr.size();i++){
             if(!ans.empty() && ans.back()[1] >= arr[i][0]){
                ans.back()[1] = max(arr[i][1],ans.back()[1]);
             }
            else{
                ans.push_back({arr[i][0],arr[i][1]});
            }
        }
        return ans;

        //TC = O(N*LognN) sorting array + O(N)  Traversal
    }
};

File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
     static int comp(int[] a, int[] b){
       if(a[0] < b[0]) return 1;
       return 0;
    }
    public int[][] merge(int[][] intervals) {
         Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        int st = intervals[0][0];
        int end = intervals[0][1];
        ArrayList<int[]> ans = new ArrayList<int[]>();
        for(int i=1;i<intervals.length;i++){
          if(end < intervals[i][0]){
            ans.add(new int[]{st,end});
            st = intervals[i][0];
            end = intervals[i][1];
          }
          else{
            end = Math.max(end , intervals[i][1]);
          }
        }

         ans.add(new int[]{st,end});
         int[][] ans1 = new int[ans.size()][2];
       for(int i=0;i<ans.size();i++){
        ans1[i][0] = ans.get(i)[0];
        ans1[i][1] = ans.get(i)[1];
       }
         return ans1;
    }
}